Blog post: First Release (#41)
rust-analyzer/rust-analyzer.github.io: blog/_posts/2020-04-20-first-release.adoc

* We try to modularize existing `rustc` compiler and extract *production ready* components for sharing with `rust-analyzer`.

For the users, the most immediately relevant facet is the first one -- a language server you can install to get smart code completion in ~~Emacs~~ your favorite editor.
This is what this post focuses on.

== What is its relationship with RLS?

Rust had a language server for quite some time now -- the https://github.com/rust-lang/rls[RLS].
RLS and rust-analyzer use fundamentally different architecture for understanding Rust.
RLS works by running a compiler on the whole project and dumping a huge JSON file with facts derived during compilation process.
rust-analyzer works by maintaining a persistent compiler process, which is able to analyse code on-demand  as it changes.
Concretely, after every keystroke RLS looks at every function body and re-typechecks it;  rust-analyzer generally processes only the code in the currently opened file(s), reusing name resolution results if possible.

rust-analyzer started as an experiment and a proof-of-concept, but today it is becoming increasingly clear that:

> * rust-analyzer already provides better experience than RLS for many users.

Reviewer: ```suggestion
* rust-analyzer already provides a better experience than RLS for many users.
```

To add rust-analyzer to VS Code:

* Remove existing rls extension, if you have one.
* Install https://marketplace.visualstudio.com/items?itemName=matklad.rust-analyzer[rust-analyzer extension] from the marketplace.

To check that everything is working open a "Hello World" Rust application.
You should see the `Run | Debug` code lens, and editor symbols should show the main function:

image::https://user-images.githubusercontent.com/1711539/80090876-7b49a500-8560-11ea-8abc-b4b5f786c026.png[]

== Features

Now that rust-analyzer is successfully installed, what are some of the most important features?

I suggest, first and foremost, to familiarize oneself with many *navigation* capabilities, as we spend more time reading code than writing it.
Here's an inexhaustive list of features.

Definition kbd:[F12]::
The most famous navigation shortcut.
One rust-analyzer specific trick is that kbd:[F12] on an `mod submodule;` brings you to the `submodule.rs` file.
This is useful in combination with:

Parent module (no default keybinding)::
This action brings to the `mod` declaration which declared the current module.

> It doesn't have a shortcut assigned by default, as there's no corresponding build-in action, but it is highly recommended to assign one.

Reviewer: ```suggestion
It doesn't have a shortcut assigned by default, as there's no corresponding built-in action, but it is highly recommended to assign one.
```

Workspace Symbol kbd:[Ctrl+T]::
This is probably the shortcut I use most often.
It is a fuzzy-search interface for all "symbols" (structs, enums, functions, field) in the project, its dependencies and the standard library.
The search tries to be smart, in that, by default, it looks only for types in your project, and, failing that, for functions.
It is possible to force search in dependencies by adding `#` to the query string, and search for all symbols by adding `*`.
Unfortunately, this doesn't work in VS Code at the moment, as it stopped passing these symbols to the language server since the last update.

video::/assets/blog/first-release/workspace-symbol.webm[options="autoplay,loop",width=100%]

Document Symbol kbd:[Ctrl+Shift+O]::
Like workspace symbol, but for things in the current file.
The same underlying LSP request powers file outline and breadcrumbs.

image::https://user-images.githubusercontent.com/1711539/80090645-1e4def00-8560-11ea-901d-d1cdc0ab8f50.png[]

Implementation kbd:[Ctrl+F12]::

> This shortcut works on structs, enums and traits, and will should you the list of corresponding impls.

Reviewer: ```suggestion
This shortcut works on structs, enums and traits, and will show you the list of corresponding impls.
```

This action runs the test function, test module or main function at the given cursor position.
It is also available as a code-lens, but I personally exclusively use kbd:[ctrl+r] for it, as I need this action all the time.
What's more, with a short cut you can re-run the last command, which is hugely useful when you are debugging a failing test.
This action is pretty smart in that it does the following things for you:

* determines the appropriate `--package` argument for `Cargo`,
* uses the full path to the test, including the module,
* sets the `--no-capture` argument, so that debug prints are visible,
* sets the `RUST_BACKTRACE` environmental variable, so that you don't have to re-run on panic.

Sadly, such context-dependent run configurations are not a part of the LSP protocol yet, so this feature is implemented using a custom protocol extension.

video::/assets/blog/first-release/run.webm[options="autoplay,loop",width=100%]

Punctuation-aware code completion::
I sometimes write code as well, and rust-analyzer has a couple of feature in this department.
When completing `return`, rust-analyzer checks if the return type is `()`.

> When completion function and method calls, `rust-analyzer` places the cursor between parenthesis, unless the function has zero arguments.

Reviewer: ```suggestion
When completing function and method calls, `rust-analyzer` places the cursor between parentheses, unless the function has zero arguments.
```

When typing `let`, rust-analyzer tries to helpfully add the semicolon.

video::/assets/blog/first-release/return-completion.webm[options="autoplay,loop",width=100%]

Extend selection kbd:[Shift+Alt+->]::

This is again a feature which is relatively simple to implement, but a huge helper.
It progressively selects larger and larger expressions, statements and items.
It works exceptionally well in combination with multiple cursor.

> One hidden capability of this feature is a navigation help: if you are in a middle of a function, you can get to the beginning of it by extending seleciton several times, and then pressing kbd:[<-].

Reviewer: ```suggestion
One hidden capability of this feature is a navigation help: if you are in a middle of a function, you can get to the beginning of it by extending selection several times, and then pressing kbd:[<-].
```

video::/assets/blog/first-release/extend-selection.webm[options="autoplay,loop",width=100%]

Fixit for missing module::

Another disproportionally nice feature -- to create a new file, type `mod file_name;` and use kbd:[ctrl+.] to add the file itself.

video::/assets/blog/first-release/new-mod.webm[options="autoplay,loop",width=100%]

Assists::

More generally, there are a lot of cases where the light bulb can write some boring code for you.
Some of my favorites are impl generation:

video::/assets/blog/first-release/impl-iter.webm[options="autoplay,loop",width=100%]

And filling match arms:

video::/assets/blog/first-release/fill-match-arms.webm[options="autoplay,loop",width=100%]

== Drawbacks

Of course, rust-analyzer being a young tool, it comes with a lot of limitations.
The most significant one is that we are not at the moment using `rustc` directly, so our capabilities for detecting errors are limited.

In particular, to show inline errors we are doing what Emacs has been doing for ages -- running `cargo check` after the file is saved.

> if the auto-save is enabled in the editor, the result is actually quite nice for small projects.

Reviewer: ```suggestion
If auto-save is enabled in the editor, the result is actually quite nice for small projects.
```

> For bigger projects though, I feel like `cargo check` in background gets in a way.

Reviewer: ```suggestion
For bigger projects though, I feel like `cargo check` in background gets in the way.
```

> So for `rust-analyzer` I have `rust-analyzer.checkOnSave.enabled = false;` in settings.

Reviewer: ```suggestion
So for `rust-analyzer` I have `rust-analyzer.checkOnSave.enabled = false;` in the settings.
```

Instead, I use the **Run** functionality to run `check` / `test` and keyboard shortcuts to navigate between errors.

Another big issue is that at the moment we, for simplicity, don't persist caches to disk.

> That means that every time you open a new project with rust-analyzer, it needs to analyzer, from source:

Reviewer: ```suggestion
That means that every time you open a project with rust-analyzer, it needs to analyze, from source:
```

* all sysroot crates (std, core, alloc, etc)
* all crates.io dependencies
* all crates in your workspace

This takes time, tens of seconds for medium sized projects.

Similarly, because we never save anything to disk, we need to keep analysis results for all crates in memory.

> At the moment, rust-analyzer process requires gigabytes of ram.

Reviewer: ```suggestion
At the moment, the rust-analyzer process can require gigabytes of ram.
```

For small projects it doesn't :grimacing:

> Finally, because analysis is not complete, features are not working correctly every time.

Reviewer: ```suggestion
Finally, because analysis is not complete, some features do not work correctly every time.
```

Sometimes there are missing completions, sometimes goto definition is wrong, we may even show false-positive errors on occasion.

This is an alpha release.
We have a long road ahead of us towards solid and reliable IDE support.

> Luckily (and this is the instance where a life of IDE writer is simpler than that of compiler writer) and IDE doesn't have to be 100% correct to be useful.

Reviewer: ```suggestion
Luckily (and this is the instance where a life of an IDE writer is simpler than that of a compiler writer), an IDE doesn't have to be 100% correct to be useful.
```

== How can I help?

If you find rust-analyzer useful and you use it professionally, consider asking your company to sponsor rust-analyzer.
We accept sponsorship at https://opencollective.com/rust-analyzer/.
You can also write directly to office@ferrous-systems.com.

> We, of course, are also open to individual backers, but we'd love to focus on corporate sponsorship.

Reviewer: maybe also mention that people can contribute code? ;)

Author: Yeah, I've run out of the fuel by the end, I hope to revisit this section before Monday. I am hesitant a bit about direct call for participation, as I don't feel good about our current review capacity...

Reviewer 2: "but we'd love to focus on corporate sponsorship." — I would name the reason openly: processing costs make corporate sponsorships more efficient (or smth along these lines)

image::https://user-images.githubusercontent.com/1711539/80091615-b5677680-8561-11ea-82de-e1517e4fef18.png[]

Run (no default keybinding)::
After navigation, the feature I use most is probably the **Run** button.
This action runs the test function, test module or main function at the given cursor position.
It is also available as a code-lens, but I personally exclusively use kbd:[ctrl+r] for it, as I need this action all the time.
What's more, with a short cut you can re-run the last command, which is hugely useful when you are debugging a failing test.
This action is pretty smart in that it does the following things for you:

* determines the appropriate `--package` argument for `Cargo`,
* uses the full path to the test, including the module,
* sets the `--no-capture` argument, so that debug prints are visible,
* sets the `RUST_BACKTRACE` environmental variable, so that you don't have to re-run on panic.

Sadly, such context-dependent run configurations are not a part of the LSP protocol yet, so this feature is implemented using a custom protocol extension.

video::/assets/blog/first-release/run.webm[options="autoplay,loop",width=100%]

Punctuation-aware code completion::

> I sometimes write code as well, and rust-analyzer has a couple of feature in this department.

Reviewer: ```suggestion
I sometimes write code as well, and rust-analyzer has a couple of features in this department.
```

Author: gosh, I've botched a rebase, should be better now
